### ISSUE TITLE
Pipeline UI does not differentiate between an artifact in GCS not existing and being unauthorized

### ISSUE THREAD

**Issue by mattnworb** [kubeflow/pipelines#1867]
**What happened:**
A pipeline component which uses `.use_gcp_secret(..)` to mount a Secret containing a GCP Service Account key has an output artifact (visualization) written to a GCS bucket. The container running the component mounts the non-Kubeflow-SA service account just fine and is able to write to the bucket successfully.

The ml-pipeline-ui application, when it tries to render this artifact, fails to access it because it's SA does not have access, and just shows a blank screen:

![image](https://user-images.githubusercontent.com/113840/63193163-bfcc8a80-c03a-11e9-9779-95df9a21213f.png)

```
GET /pipeline/artifacts/get?source=gcs&bucket=...&key=...
Getting storage artifact at: gcs: .../output_display.html
No matching files found.
```

I think this is because the ml-pipeline-ui frontend server code, when trying to retrieve files from a GCS bucket in doesn't pass a callback to `bucket(bucket).getFiles(...)`:

https://github.com/kubeflow/pipelines/blob/a16f3806374034b8f9898fe33c60eba46c5ae282/frontend/server/server.ts#L154

Looks like the `@google-cloud/storage` library [will signal an error in loading the files from GCS by invoking the callback passed to `getFiles(...)`](https://github.com/googleapis/nodejs-storage/blob/v2.5.0/src/bucket.ts#L1953) - but the kubeflow/pipelines code here does not pass any callback. (I am not familiar with Typescript so I might be interpreting the code wrong)

**What did you expect to happen:**
If ml-pipeline-ui can't load the artifact because it is not authorized, to display an error rather than a blank window.

**Anything else you would like to add:**
This is a different issue than https://github.com/kubeflow/pipelines/issues/1568 although the title matches my description, since in this case the component has access to write to the GCS bucket - it is just that the UI can't seem to tell the difference between not being able to load the artifact from GCS because it doesn't exist versus being unauthorized.

**Comment by Bobgy** [comment]
The specific usage isn't wrong. We are using promise to get result instead of callback. That's expected.

However, whether the client could distinguish between an object doesn't exist or there is no permission is something I don't know. Needs some investigation on this.

Please let me know if you know more about this.

**Comment by mattnworb** [comment]
@Bobgy I might be reading this Typescript code wrong, but it looks like`@google-cloud/storage` doesn't return a `Promise` itself - it has `return fileInstance`. Errors are signalled by invoking the callback parameter, if one is passed. I believe this means that the code invoking this function by using`await` will only get "fulfilled" results, and not rejections (when an error is returned / thrown).

https://github.com/googleapis/nodejs-storage/blob/v2.5.0/src/bucket.ts#L1950-L1972

**Comment by Bobgy** [comment]
I wrote a simple script to test the behavior:
```
// test.js
async function test() {
  const { Storage } = require('@google-cloud/storage');
  const storage = new Storage();
  try {
    const files = await storage.bucket('abcdelksajldkjsldfjls').getFiles({ prefix: '' }); // this bucket doesn't exist
    console.log(files[0][0].name);
  } catch (err) {
    console.log('' + err);
    if (err.code === 404) {
      console.error('Not found');
    }
  }
  try {
    const files = await storage.bucket('abcde').getFiles({ prefix: '' }); // this bucket exists, but I don't have permission
    console.log(files[0][0].name);
  } catch (err) {
    console.log('' + err);
    if (err.code === 403) {
      console.error('No permission');
    }
  }
}

test();
```

Result:
```
$ node test.js
Error: Not Found
Not found
Error: gongyuan@google.com does not have storage.objects.list access to abcde.
No permission
```

We can use different promise rejection error codes to distinguish "Not found" and "No permission", at least at bucket level.

**Comment by Bobgy** [comment]
I tried again with per file permission using an account with only object list permission and got the following result:
```
// test.js

async function print(file) {
  if (file) {
    const [buffer] = await file.download();
    console.log(buffer.toString().substring(0, 40));
  }
}

async function testFileLevel() {
  const { Storage } = require('@google-cloud/storage');
  const storage = new Storage();
  try {
    const [files] = await storage.bucket('gongyuan-test2').getFiles({ prefix: 'does-not-exist' });
    console.log(files.length);
    await print(files[0])
  } catch (err) {
    console.log('' + err);
    if (err.code === 404) {
      console.error('Not found');
    }
  }
  try {
    const [files] = await storage.bucket('gongyuan-test2').getFiles({ prefix: 'tfx_sample_fix.py' });
    console.log(files.length);
    await print(files[0])
  } catch (err) {
    console.log('' + err);
    if (err.code === 403) {
      console.error('No permission');
    }
  }
}

testFileLevel();
```

Results:
```
$ node test.js
0
1
Error: gongyuan94@gmail.com does not have storage.objects.get access to gongyuan-test2/tfx_sample_fix.py.
```

It seems this code also can distinguish between not exist and no permission.

**Comment by Bobgy** [comment]
It's strange, then why does our deployment show there are no matched files when there is actually an error?

I'm investigating

**Comment by Bobgy** [comment]
Strange...
I tried ml-pipeline-ui:0.1.40, reproduced the same problem as this issue. When my UI pod has no permission, the error message was 'No files matched'.

However, if I tried ml-pipeline-ui:0.2.0, this is no longer reproduceable. I got a reasonable error message:
```
Failed to download GCS file(s). Error: Error: Primary: /namespaces/gongyuan-pipeline-test.svc.id.goog with additional claims does not have storage.objects.list access to gongyuan-test.
``

**Comment by Bobgy** [comment]
The only change that could have affected this: https://github.com/kubeflow/pipelines/pull/2830

**Comment by Bobgy** [comment]
```
$ nvm use 12
Now using node v12.3.1 (npm v6.12.0)
$ node test.js
Error: Not Found
Not found
Error: gongyuan@google.com does not have storage.objects.list access to abcde.
No permission
$ nvm use 9.4.0    
Now using node v9.4.0 (npm v5.6.0)
$ node test.js 
0
0
```
Verified using my first bucket level test script. It's indeed the node version difference.

/close
So this issue has already been fixed by #2830.

**Comment by mattnworb** [comment]
@Bobgy interesting, thanks for digging into this!
